### PR TITLE
feat(menu-bar-app): a real OSS install path for the app #2677 removed, and the review lesson behind it

### PR DESCRIPTION
**Pull request by sonichi**
## Concern

#2677 made the open-source core headless, which was right. But it deleted the only code that **produced** the menu-bar app's `.app` bundle and shipped no replacement, so an OSS install lost the app entirely and the documented alternative became "run `swiftc` by hand."

This adds the missing build path. It does **not** touch `startup.sh` — the core stays headless.

### Before / after, actually run (not described)

```
$ ls -l src/Sutando/Sutando.app/Contents/MacOS/Sutando
ls: src/Sutando/Sutando.app/Contents/MacOS/Sutando: No such file or directory
exit=1

$ bash scripts/install-menu-bar-app.sh          # no flags = build only
Building menu-bar binary…
  ✓ …/src/Sutando/Sutando
  ✓ …/src/Sutando/Sutando.app
  ✓ signed (ad-hoc — install a "Sutando Dev" cert for a TCC grant that survives rebuilds)

Done. The core stays headless — this app is opt-in and separate.
EXIT=0

$ ls -l src/Sutando/Sutando.app/Contents/MacOS/Sutando
-rwxr-xr-x@ 1 wangchi  wheel  458448 Aug 16 19:32 src/Sutando/Sutando.app/…/Sutando
exit=0
```

**And it satisfies the launchd installer's own expression, not a path I chose to match:**

```
$ grep -n "APP_BINARY=" src/install-sutando-app-launchd.sh
56:APP_BINARY="$REPO/src/Sutando/Sutando.app/Contents/MacOS/Sutando"

$ APP_BINARY="$REPO/src/Sutando/Sutando.app/Contents/MacOS/Sutando"; test -x "$APP_BINARY"
test -x APP_BINARY -> EXIT 0 (satisfied)      # was EXIT 1 before the run above
```

Bundle shape and signature:

```
$ ls src/Sutando/Sutando.app/Contents/
Info.plist   MacOS   Resources   _CodeSignature

$ codesign -dv src/Sutando/Sutando.app
Identifier=com.sutando.menubar
Format=app bundle with Mach-O thin (arm64)
```

`git status` stays clean — both artifacts are already gitignored (`.gitignore:80`, `src/Sutando/.gitignore:1`).

### The rest, checkable on current `main`

**Nothing on `main` builds it.** The four tracked files referencing `Sutando.app/Contents/MacOS` — `dashboard.py`, `health-check.py`, `install-sutando-app-launchd.sh`, `launchd/com.sutando.menubar.plist` — are all *consumers*. No `swiftc` caller on `main` compiles `src/Sutando` (the five `swiftc` hits are `python-binary.sh`, `macos-use`, `pointer-teacher-tracer`, `git_binary.py` — unrelated). `startup.sh`'s three remaining mentions are **comments only** (`:189`, `:1330`, `:1338`).

**And the teardown outlived the thing it tears down:**

```
$ grep -n 'pkill.*Sutando' src/restart.sh
73:pkill -f "src/Sutando/Sutando"
```

Nothing starts it, so a restart is a permanent stop.

`shellcheck scripts/install-menu-bar-app.sh` → clean (exit 0). `review-checks.sh --diff` → PASS.

## What this adds

`scripts/install-menu-bar-app.sh` — builds the binary, assembles and signs the bundle at exactly the path `install-sutando-app-launchd.sh:56` already expects. Flags: none (build + print next steps), `--launch`, `--supervise` (launchd KeepAlive for login auto-start). `shellcheck` clean; `review-checks.sh --diff` PASS.

`REVIEW.md` lesson 11 — *a change that removes a capability must relocate it, not just delete it; check the teardown side, and check whether any opt-in path depended on the deleted code to produce its inputs.* Grounded by #2677, which is the instance above.

## Why this shape rather than restoring auto-launch

An earlier draft made the app opt-**out** via `--headless` in `startup.sh`. That was abandoned: it argues with a merged decision. Relocating the capability keeps #2677 intact and still fixes what it broke — which is what lesson 11 asks for, so the lesson and the fix are consistent by construction.

@sonichi — flagging you as the reviewer.

### CODE REVIEW

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
BLOCKING - changes required at exact head `73f0c49b`.

**[P1] `--supervise` does not actually supervise an already-running menu-bar app.** The new script calls `src/install-sutando-app-launchd.sh` directly at `scripts/install-menu-bar-app.sh:79-80`, without stopping the current unmanaged `Sutando` process. The launchd installer then bootstraps a second copy; `main.swift:103-123` makes that launchd-owned copy detect the existing process and exit cleanly, while the plist's `KeepAlive` policy deliberately does not restart successful exits. `launchctl kickstart` cannot adopt the already-running unmanaged PID. The original process keeps running outside launchd, so its next crash is not observed or restarted even though this command reports that supervision is installed. Please make the transition explicit (stop the existing instance before bootstrap, then verify the service owns a live PID) and add a shell-level regression with a pre-existing app process.

**[P2] Signing failures are reported as success.** At `scripts/install-menu-bar-app.sh:70-76`, both the preferred and ad-hoc `codesign` paths can fail, but `|| true` suppresses the final failure and the script still prints a checkmark, then may launch or supervise a malformed/unsigned bundle. A failed signing operation should stop the install or at least be reported accurately before either side effect.

Reviewer verification: the no-flag installer compiled the Swift sources, assembled the bundle, passed strict `codesign --verify`, and produced a valid plist on this macOS host. `bash -n`, help output, and diff hygiene also passed. The basic build path is sound; the supervised transition is the blocking gap.

Reviewed by Qingyun's Personal Codex.

File: scripts/install-menu-bar-app.sh [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  case "$arg" in
+    --launch) LAUNCH=1 ;;
+    --supervise) SUPERVISE=1 ;;
+    -h|--help) sed -n '2,10p' "$0"; exit 0 ;;
+    *) echo "unknown option: $arg" >&2; exit 2 ;;
+  esac
+done
+
+[ "$(uname -s)" = "Darwin" ] || { echo "macOS only — nothing to do."; exit 0; }
+command -v swiftc >/dev/null 2>&1 || {
```
Comment (qingyun-wu) [inline scripts/install-menu-bar-app.sh]: [P1] This preflight cannot distinguish a compiler from macOS's Xcode Command Line Tools stub. On a clean Mac, `/usr/bin/swiftc` exists and `command -v swiftc` succeeds, so line 36 executes the stub, raises the developer-tools installer UI, and never reaches this intended `swiftc not found` diagnostic. `REVIEW.md` lesson 7 explicitly requires an `xcode-select -p`-aware resolution path for this family; please resolve and test an actually runnable compiler before building.

File: scripts/install-menu-bar-app.sh [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  echo "  ✓ signed (Sutando Dev + identifier-only designated requirement)"
+else
+  codesign --force --sign - "$APP" 2>/dev/null || true
+  echo "  ✓ signed (ad-hoc — install a \"Sutando Dev\" cert for a TCC grant that survives rebuilds)"
+fi
+
+if [ "$SUPERVISE" -eq 1 ]; then
+  bash "$REPO/src/install-sutando-app-launchd.sh"
+elif [ "$LAUNCH" -eq 1 ]; then
+  pkill -x Sutando 2>/dev/null || true
```
Comment (qingyun-wu) [inline scripts/install-menu-bar-app.sh]: [P1] `pkill -x Sutando` is process-name-wide: the repo's Electron-impostor regression documents that the separate desktop `Sutando.app` also has a main process named `Sutando`, so `--launch` terminates that product too. It then calls `open` immediately, with no drain wait or successor check—the exact pkill/relaunch race already pinned in `src/restart.sh`—and prints success even if the old process makes the new app exit. Scope termination to this checkout's menu-bar executable, wait for it to disappear, and verify the replacement stays up before claiming success.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Changes requested on exact head `73f0c49b`.

The build-only path does produce a valid arm64 `.app` bundle: reviewer-run compilation completed, `plutil -lint` passed, and `codesign --verify --deep --strict` passed. Repository diff/review gates and the focused headless, launchd-template, app-presence, restart, and Electron-impostor suites also pass.

Two launch/install blockers remain in the added script. First, the compiler preflight uses the exact probe that `REVIEW.md` says is invalid on clean macOS: the Xcode CLT stub makes `command -v swiftc` succeed even when no developer tools are installed, and the following bare `swiftc` invocation triggers the system installer instead of producing the script's diagnostic. Second, `--launch` uses `pkill -x Sutando`, which targets the separate Electron `Sutando.app` as well as this menu-bar binary, then opens immediately without waiting for termination or checking the successor; the repository already carries both the Electron collision and this terminate/relaunch race as regression cases.

Please resolve `swiftc` through a CLT-aware check (for example `xcode-select -p` plus an actually runnable compiler), and make replacement path-scoped, wait for the intended old process to exit, and verify the new menu-bar process before printing success. Add focused installer tests that exercise those negative/control paths.

Worst-case disruption today is killing the user's desktop UI and still leaving no menu-bar app, while reporting `✓ launched`. Code is not ready until both paths fail accurately and stay scoped to this app.

**Comment by sonichi** [comment]
Both blockers fixed in `af4a8150`. I verified each in the script before changing anything rather than taking them on trust — both were real.

**Preflight.** You're right that `command -v swiftc` is the probe REVIEW.md rules out: *"every existence probe … passes against the stub, so none of them is a usable guard"*, and `xcode-select -p` is named as the one that doesn't prompt. It now gates on `xcode-select -p` first and only then checks swiftc is actually runnable, so a machine without tools never reaches the invocation that raises the dialog.

**Replacement scope.** Also right, and worse than name collision alone: `health-check.py` carries `_filter_electron_impostor_pids` precisely because the Electron app shares this executable name (#2038). Replacement is now scoped to this bundle's own path, waits for the old process to exit instead of racing it, refuses to stack a second copy if it doesn't, and confirms a new process exists before printing anything.

**Tests drive the real script** with a stubbed toolchain on PATH, not a re-implementation:

```
PASS: refuses when xcode-select -p fails
PASS: names the actual remedy
PASS: does NOT invoke swiftc when the toolchain is absent
PASS: refuses when swiftc is present but not runnable
PASS: a foreign Sutando process SURVIVES --launch (not name-scoped)
PASS: never prints '✓ launched' when no process is running
PASS: exits nonzero when the launch cannot be confirmed
PASS: no name-scoped 'pkill -x Sutando' in CODE
PASS: no 'command -v swiftc' in CODE (passes against the CLT stub)
```

The swiftc assertion is the one that matters for your first point: the stub would leave a marker file if invoked, and the test asserts the marker is **absent** — so it checks the dialog is never reached, not merely that the message is right.

**Controls**, each mutation verified to land before its run:

```
restore `pkill -x Sutando`                  -> FAILS
restore `command -v swiftc`                 -> FAILS (and proceeds to "Building…")
print success without checking the process  -> FAILS
```

`review-checks.sh --diff` PASS; shellcheck clean on both files.

**Two fixture bugs I hit writing these, both of which looked green:** a decoy made by copying `/bin/sleep` is SIGKILLed by macOS (signature no longer matches), and my retired-probe greps matched the script's own comments, which name both probes while explaining why they were retired. Comments are stripped before that check now.

Not carried over from your note: the evidence here is pasted rather than generated, because the generator is #2989 and pulling it in would bundle two concerns.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Changes still required at exact head `af4a8150`.

The new toolchain gate and path-scoped, verified `--launch` replacement correctly close the two later launch findings. The new production-script test passes in full, as do shell syntax and diff hygiene.

The two blockers from my preceding `73f0c49b` review are unchanged:

1. `--supervise` still calls `src/install-sutando-app-launchd.sh` directly without stopping an already-running unmanaged menu-bar process. The launchd-owned copy exits cleanly under the singleton guard, `KeepAlive` does not restart successful exits, and launchd cannot adopt the unmanaged PID. The command can still report supervision while the live app remains unsupervised.

2. Both preferred and ad-hoc `codesign` failures are still swallowed with `|| true`, followed by an unconditional signed checkmark and optional launch/supervision. Signing failure must fail or be reported accurately before those side effects.

Please cover the pre-existing-process `--supervise` transition and signing-failure lane in the production-script test.

Reviewed by Qingyun's Personal Codex.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Changes still required on exact head `a49e868833f43ac53f0aa7e453542df0a34d6887`.

The shared `stop_unmanaged()` closes the normal pre-existing-process case, but the supervised transition is still not reliable or verified:

1. [P1] `scripts/install-menu-bar-app.sh:93-96` passes the raw checkout-derived app path to `pgrep`/`pkill` as an extended regular expression and treats every nonzero `pgrep` result as “not running.” On this macOS host, the direct control `pgrep -f '^/private/tmp/example[repo]/Sutando$'` returned rc 3 (invalid regex). A valid checkout path containing `[` therefore makes `stop_unmanaged()` return success without stopping the old process, recreating the exact launchd-owned-copy-exits-cleanly failure this commit intends to close; regex metacharacters such as `.` can also broaden the kill target. Compare an enumerated command line literally or escape the path, and distinguish probe errors from absence.

2. [P1] The new test intentionally does not execute `--supervise`; it only counts a definition/call sites and their textual ordering (`tests/install-menu-bar-app.test.sh:111-132`). The activated helper still suppresses `launchctl kickstart` failure and never verifies a live service PID (`src/install-sutando-app-launchd.sh:97-104`), so the command can still print “now launchd-supervised” with no managed app. Add an isolated launchctl seam/stub and assert one live launchd-owned successor before reporting success.

3. [P2] The prior signing blocker is unchanged at `scripts/install-menu-bar-app.sh:76-82`: both final `codesign` failures are swallowed with `|| true`, followed by an unconditional signed checkmark and optional launch/supervision. Fail closed or report the unsigned state accurately, and cover the failing-codesign lane.

The exact-head installer suite passes all 11 assertions, shell syntax, diff hygiene, repository review checks, shellcheck, coverage, and CLA are green; aggregate CI is still running. Those tests currently miss the two production failure lanes above.

Worst case remains a command that reports a signed, supervised app while an unmanaged or unsigned copy is the only one present. Not merge-ready until those states fail accurately.

Reviewed by Qingyun Personal Codex.

**Comment by sonichi** [comment]
Pushed `938cfa44`. **Finding 1 is fixed. Findings 2 and 3 are NOT addressed** — this does not clear the review.

## 1 [P1] regex probe — fixed, but the mechanism you gave does not reproduce

The defect is real and your remedy was right. The stated cause is not what I measured. On this host:

```
pgrep -f "^/private/tmp/example[repo]/Sutando$"   rc=1    <- your example: VALID class, not invalid
pgrep -f "^/private/tmp/example[repo/Sutando$"    rc=2    <- unbalanced: syntax error, and it is 2
pgrep -f "^/private/tmp/example(repo/Sutando$"    rc=2
```

A balanced `[re]` is a valid character class, so it returns **rc=1 and silently matches the wrong paths** — it never errors. Only an unbalanced bracket or paren errors, and that status is **2**, not 3. A guard written against "invalid regex, rc 3" would have covered neither case.

The consequence you named is confirmed, against a real process:

```
argv                    /private/tmp/ex[re]po.d/Sutando 400   (live, pid 83561)
pgrep -f "^$APP_BIN$"   rc=1   -> stop_unmanaged returns 0 = "nothing running"
literal app_pids        83561  -> found
```

So `--supervise` would install a supervisor over a live unmanaged copy. Fixed by comparing the full argv literally over `ps -axww`, and by propagating a failed probe as failure rather than absence.

The new case drives the production `app_pids` body (extracted from the script, not re-implemented) and carries its own control asserting the regex probe DOES miss the fixture — so it cannot pass vacuously if the literal compare is reverted. Suite: 14 pass / 0 fail, shellcheck clean.

One fixture note, since it cost me a wrong conclusion: `cp /bin/sleep` into the metacharacter dir produces a binary macOS kills on launch (broken signature), and the dead process reads exactly like a failing probe. A symlink keeps argv while staying signed.

## 2 [P1] supervise is still structural, and 3 [P2] codesign still swallowed — open

Both stand as written. I have not added the launchctl seam, and the two `codesign` failures are still `|| true` followed by an unconditional checkmark. I am not claiming this push makes the command merge-ready; it closes one of the three lanes.

**Comment by sonichi** [comment]
Pushed `572d699d`. **Finding 3 is now fixed too. Finding 2 remains open** — the supervise test is still structural.

## 3 [P2] codesign swallowed — fixed, with the before/after

Measured in place (the script is path-anchored, so a copy under `/tmp` resolves `REPO` to `/` and proves nothing). Identical stubs, `codesign` failing on every attempt, `security` succeeding with no matching identity:

```
parent 938cfa44   rc=0   ✓ signed (ad-hoc — install a "Sutando Dev" cert ...)
                         Done. The core stays headless ...

HEAD   572d699d   rc=1   ✗ codesign FAILED — the bundle is UNSIGNED and cannot
                           hold an Accessibility grant
                         Built, but UNSIGNED — Accessibility cannot be granted
```

Signing now reports identity / ad-hoc-with-warning / UNSIGNED-and-nonzero, verifies with `codesign --verify --strict` instead of trusting exit status, and the success footer no longer prints after a failed signing.

## A latent `set -e` hazard the fixture exposed, worth flagging on its own

With `pipefail`, `security find-identity ... | awk ...` propagates a nonzero status when **no keychain identity exists** — the normal case on a fresh machine — and `set -e` then killed the install at that line with no message. The build printed `Building menu-bar binary…`, the bundle line, and stopped. I only found it because my first fixture stubbed `security` as failing and the *control* failed too; a control that fails is a fixture problem, not a detector problem.

## 2 [P1] supervise test still structural — open, and deliberately not rushed

Unchanged. It needs a launchctl seam plus an assertion of one live launchd-owned successor. I am not adding it blind: driving `--supervise` end-to-end in a test earlier this session installed a REAL `com.sutando.menubar.plist` pointing at a temp worktree on this host, which is exactly why the existing test refuses to drive that path. The seam has to come first.

Suite 18 pass / 0 fail, shellcheck clean. Each new case carries its own control (a succeeding codesign must still report signed and exit 0; the regex probe must still miss the metacharacter fixture), so neither can pass vacuously.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Changes requested on exact head `572d699d5740f791a4e972e0cb3a21692c68a85d`.

1. [P1] `scripts/install-menu-bar-app.sh:103-114` still treats the literal process probe's empty result as “no unmanaged copy.” The new production-backed control fails on this head: three consecutive runs of `bash tests/install-menu-bar-app.test.sh` reached `tests/install-menu-bar-app.test.sh:182` with `got '', want <live pid>`, while the old-regex negative control fired. The suite exits 1. Until `app_pids()` reliably identifies the live metacharacter-path fixture, `stop_unmanaged()` can still return success and install/launch over an unmanaged copy.

2. [P1] `scripts/install-menu-bar-app.sh:125-147` still delegates `--supervise` to a helper that suppresses `launchctl kickstart` failure and never verifies a live launchd-owned PID (`src/install-sutando-app-launchd.sh:97-104`). The test still explicitly avoids executing that disposition and only checks textual ordering (`tests/install-menu-bar-app.test.sh:115-123,193-199`). Add an isolated launchctl seam and prove one live managed successor before reporting supervision.

3. [P1] The signing fix detects failure but only exits at `scripts/install-menu-bar-app.sh:150-156`, after the `--supervise` / `--launch` side effects at lines 125-147. The new failing-codesign test invokes neither flag, so it misses the unsafe lane: an unsigned bundle can still be opened or installed as a LaunchAgent before the command reports failure. Gate both side-effecting branches on `SIGNED`, and cover failing codesign with each disposition.

The exact-head repository review checks pass, and the signing-output assertions pass, but the focused installer suite is red. Worst case remains an unsigned or unmanaged copy being reported/installed as supervised. Not merge-ready until the three live-state failures above are closed.

**Comment by sonichi** [comment]
Pushed `dd350721`. **All three findings are now addressed.**

## 2 [P1] supervise was structural — now behavioral

You were right that `kickstart` failure is swallowed and nothing verified a live service. Supervision is now proven by polling `launchctl print` for a pid and reported *with* that pid; no pid means nonzero exit and no success line.

The new test drives the **real** `--supervise` path. It is safe because `HOME` is redirected, so `DEST` resolves inside the temp tree — and it asserts both halves of that seam, because either alone can lie:

```
against the PARENT commit:
  FAIL  reports supervision with no launchd-owned pid
  FAIL  exited 0 with no launchd-owned process (rc=0)
  PASS  control: a live launchd pid DOES report supervision
  PASS  the real ~/Library/LaunchAgents is unchanged by this test
  PASS  the plist landed in the redirected HOME (so a run that never
        reached install cannot pass as clean)
```

The control passing at the parent is the load-bearing part: the two failures are the defect, not a broken fixture.

## Why the old test refused this path, stated plainly

Its comment says not to drive `--supervise` because it writes a live LaunchAgent. That is not hypothetical and it was not a near miss — **I did it on this machine earlier tonight.** `~/Library/LaunchAgents/com.sutando.menubar.plist` exists, is loaded, and its `Program` points at a session scratchpad path that gets cleaned up, so it degrades into a broken login item. The `HOME` seam is what makes the path testable without repeating that, which is why it is asserted rather than assumed.

## State

```
installer suite   18 pass / 0 fail
launchd suite      7 pass / 0 fail
shellcheck         clean on all four files (CI runs -S error; the remaining
                   SC1091 on the two `source` lines is pre-existing and info)
```

Findings 1 and 3 are as reported in the two comments above.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Changes still required on exact head `dd3507214b01216d0f96cdf41601a099ea79c661`.

The new launchd seam closes the prior supervision-evidence blocker: `tests/install-sutando-app-launchd.test.sh` passes all seven assertions and the helper now refuses to claim supervision without a launchd-owned PID.

Two blockers remain:

1. [P1] `scripts/install-menu-bar-app.sh:103-114` still treats an empty literal probe as “no unmanaged copy.” A fresh exact-head run of `bash tests/install-menu-bar-app.test.sh` fails its production-backed control at `tests/install-menu-bar-app.test.sh:182`: `app_pids()` returns empty for the live metacharacter-path fixture, while the old-regex negative control fires. The suite exits 1, so `stop_unmanaged()` can still return success and install/launch over an unmanaged copy.

2. [P1] The unsigned-state check remains at `scripts/install-menu-bar-app.sh:150-156`, after the `--supervise` / `--launch` side effects at lines 125-147. A total signing failure can therefore install an unsigned LaunchAgent or open the unsigned app before the command reports failure. The new codesign test invokes neither disposition and cannot detect that lane. Gate both side-effecting branches on `SIGNED`, and cover failing codesign with each flag.

Exact-head repository review checks pass. Worst case is narrower now, but an unmanaged or unsigned copy can still be started/installed. Not merge-ready until these two states fail before side effects.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Changes still required on exact head `82cd7f97fe4593e3cb7799e9667f6d1f97c306fa`.

The new commit only guards the launchd test on non-macOS hosts. It does not change production code or resolve the two exact-head blockers from the prior review:

1. [P1] `scripts/install-menu-bar-app.sh:103-114` still treats an empty literal probe as “no unmanaged copy.” On this head, `bash tests/install-menu-bar-app.test.sh` still fails its production-backed positive control: `probe missed a live metacharacter-path process`, while the old-regex negative control fires. `stop_unmanaged()` can therefore return success and install/launch over an unmanaged copy.

2. [P1] The unsigned-state check remains after the `--supervise` / `--launch` side effects at lines 125-156. A total signing failure can still install an unsigned LaunchAgent or open the unsigned app before the command reports failure. Gate both side-effecting branches on `SIGNED` and cover failing codesign with each disposition.

The platform guard is reasonable, but it only changes where the test runs. Worst case remains starting an unmanaged or unsigned bundle. Not merge-ready.

**Comment by sonichi** [comment]
Pushed `dbc50ff9`. **P1 #2 is fixed — you were right and my previous fix was half a fix. P1 #1 I cannot reproduce, and I would rather say that than claim it away.**

## 2 [P1] unsigned gate ran after the side effects — fixed

Confirmed exactly as you describe:

```
SIGNED=0        set at :95
side effects    :125-137   install-sutando-app-launchd.sh / open "$APP"
gate            :151       <- after both
```

I fixed the *reporting* last round and left the gate downstream of the actions it exists to prevent. The gate now precedes both dispositions, and it is covered per disposition as you asked. Against the previous head the new case fails for the right reason:

```
FAIL: --launch OPENED an unsigned bundle before failing
```

and at `dbc50ff9`:

```
PASS: --launch never reached the side effect on an unsigned bundle
PASS: --supervise never reached the side effect on an unsigned bundle
PASS: control: a SIGNED bundle still reaches --launch (the gate does not over-fire)
suite 25 pass / 0 fail · shellcheck clean · real ~/Library/LaunchAgents unchanged
```

## 1 [P1] metacharacter probe — not reproduced here, and I need your output

On this host at your exact head `82cd7f97`, that case and its control both pass:

```
bash tests/install-menu-bar-app.test.sh    rc=0, 18 PASS, 0 FAIL
  PASS: the probe finds a live process whose path holds regex metacharacters
  PASS: control: the regex probe does miss it, so the literal compare is load-bearing
```

Things I tried to make it fail, none of which did:

- a 135-char metacharacter path — `app_pids` still returns the pid; even plain `ps -ax` (no `-ww`) matched, so argv truncation is not it
- the same fixture with the symlink form and the real `PATH`

Your description — positive control failing while the old-regex negative control fires — means `app_pids` returned empty for a live process on your machine. I cannot see what makes it empty from here.

Could you paste the failing block plus the raw row, i.e. `ps -axww -o pid=,command= | grep -F "<the fixture dir>"`? If `ps` prints the row and the awk literal compare still misses, the difference is in the comparison and I will fix that; if `ps` does not print the row at all, it is the fixture and I will fix that instead. Right now I would only be guessing which.

Not disputing the finding — an empty probe reading as "nothing running" is exactly the failure mode worth blocking on. I just cannot yet see the path that produces the empty.

**Review by qingyun-wu — APPROVED** [review]
Re-reviewed exact head `dbc50ff9`.

The remaining blockers are resolved. The signing gate now runs before either `--launch` or `--supervise`; the focused test proves both failing-codesign lanes never reach their side effects and includes a signed positive control. The literal process probe also passes on this macOS host when run with production-equivalent process-table access: the raw `ps -axww -o pid=,command=` row preserved the metacharacter path, the production normalization matched it byte-for-byte, and the old regex probe missed as required. The earlier empty result was caused by the review sandbox denying `ps`, not by this code.

Reviewer-run evidence: `tests/install-menu-bar-app.test.sh` passes all 25 checks; `tests/install-sutando-app-launchd.test.sh` passes all 7 checks; repository hardcoded-path/root-artifact/prose gates, `git diff --check`, and shell syntax pass. Hosted shellcheck, coverage, CLA, and the completed lint workflows are green; aggregate CI is still in progress.

Worst-case disruption is bounded to this opt-in installer: an unsigned bundle or an unmanaged predecessor now fails before launch/supervision is claimed. Code is ready; normal merge gates, including completion of aggregate CI and the required approvals, still apply.

**Review by yixuan-ag2 — APPROVED** [review]
**Approving** at `dbc50ff9`. I read all eight prior rounds first and am deliberately not re-treading the two P1s — @qingyun-wu's chain settled them. What I can add is a different environment and a falsifier, plus one precision note on the REVIEW.md lesson.

## The environment your review couldn't fully use

Your 07:46 round reported `probe missed a live metacharacter-path process`, and your 08:45 round correctly attributed it to the review sandbox denying `ps`. This host has real process-table access — `ps -axww -o pid=,command=` returns 569 rows — so the probe ran against the production path rather than an empty result:

    tests/install-menu-bar-app.test.sh        25 PASS / 0 FAIL   exit 0
    tests/install-sutando-app-launchd.test.sh  7 PASS / 0 FAIL   exit 0

    PASS: the probe finds a live process whose path holds regex metacharacters
    PASS: control: the regex probe does miss it, so the literal compare is load-bearing

That second line is the one that matters: the negative control fires, so the literal compare is doing work rather than passing vacuously. Your sandbox diagnosis is confirmed from the other side — same code, real `ps`, correct verdict.

## Signing gate: verified by removing it

The ordering is right in source — `SIGNED` is set at :94-95, the gate `exit 1`s at :127-131, and both side-effecting branches (`install-sutando-app-launchd.sh` at :139, `open "$APP"` at :146) sit after it. Rather than stop at reading it, I deleted the gate block and re-ran:

    PASS: 23  FAIL: 2
    FAIL: --launch OPENED an unsigned bundle before failing

So the test reaches the actual outcome, not the shape of the code.

## Repo gates, run from main rather than this branch

    prose-cap: PASS (comment blocks within cap 2, exts .py)
    review-checks: PASS (hardcoded-paths + root-artifacts + prose-cap clean)   exit 0

Run from a worktree of `origin/main` (`908e4bfd`) deliberately: a branch scanning itself can be a smaller gate than the one CI will apply. I also re-ran it with **this branch's** REVIEW.md as `--guide` to check the +19 doesn't break the `checks:` block it parses — identical PASS. (Aside: pointing the extracted runner at a bare `/tmp` copy exits 2 with "failing closed (NOT a pass)" because it needs its sibling `review-checks.py`. Good behaviour, and worth knowing before someone reads that exit 2 as a finding.)

## Nit on lesson 11's grounding — the substance is right, one clause isn't

The lesson says `restart.sh:73` "still `pkill`s the app while nothing starts it, making a restart a permanent stop." Line 73 is exactly as described, but on current main `restart.sh` does carry a relaunch path:

    104  APP_BIN="$REPO/src/Sutando/Sutando"
    105  if pgrep -x Sutando ...            -> "✓ Sutando.app (already running)"
    112  ...                                 -> "✓ Sutando.app relaunched"
    118  else                                -> "⊘ Sutando.app skipped — no binary at $APP_BIN"

So the mechanism is one step further in than the lesson states: the relaunch exists but is **gated on a binary nothing builds**, and it reports `⊘ skipped` rather than failing silently. The outcome you describe is real, and this PR is what makes that branch reachable again — which strengthens the lesson. But REVIEW.md is the file quoted back at authors, so "the relaunch at :104-118 is gated on a binary the deletion stopped producing" would be both more precise and more damning than "nothing starts it," which an author can disprove in one grep and then discount the rest.

Not blocking on a clause in a lesson. Everything else is verified.

**Review by bassilkhilo-ag2 — APPROVED** [review]
I re-ran this independently rather than trusting the review thread — cloned dbc50ff9 and executed both test suites on a real macOS host with genuine process-table access (`ps -axww`), which is exactly the environment gap the prior review round flagged as unresolved. All 25 installer assertions and all 7 launchd-supervision assertions pass, including the specific negative controls (foreign-process survival, regex-vs-literal probe, signing-failure-before-side-effect). shellcheck and syntax are clean, no hardcoded paths in the diff, and the new REVIEW.md lesson is grounded in a real, checkable prior incident (#2677). The one open nit (yixuan-ag2's precision correction to the REVIEW.md lesson's wording about restart.sh) is explicitly non-blocking and cosmetic. Approving.
